### PR TITLE
feat: support default provider

### PR DESCRIPTION
**Pull request by WindSpiritSR**
fixed: #222 

Signed-off-by: WindSpiritSR <simon343riley@gmail.com>

### CODE REVIEW

**Comment by hsluoyz** [comment]
@WindSpiritSR demo link

**Comment by hsluoyz** [comment]
@Kininaru @sh1luo

**Comment by WindSpiritSR** [comment]
Deployed at: https://door.leviatan.cn

File: web/src/ProviderTable.js [hunk 1]
```diff
@@ -62,6 +64,13 @@ class ProviderTable extends React.Component {
...
 
   renderTable(table) {
+    table.forEach(item => {
+      if (item.default) {
+        defaultProvider = item.name;
+        return false
+      }
+    })
+
     let columns = [
```
Comment (sh1luo) [inline web/src/ProviderTable.js:74]: A little bloated. Replaced by:
```javascript
if (table.some(item => item.default)) {
    defaultProvider = item.name;
    return false;
}
```

File: web/src/ProviderTable.js [hunk 2]
```diff
@@ -113,10 +122,38 @@ class ProviderTable extends React.Component {
           return (
             <Switch checked={text} onChange={checked => {
               this.updateField(table, index, 'canSignIn', checked);
+              if (!checked && defaultProvider === record.name) {
+                defaultProvider = null;
+                this.updateField(table, index, 'default', false);
+              }
             }} />
```
Comment (sh1luo) [inline web/src/ProviderTable.js:129]: I think that it's no need to update `defaultProvider`, which is set by application admin and can be empty.

Only need to decide whether to set `default=true` according to ``canSignUp || canSignIn``

Writing in this way will cause the logic of the partial code to be complicated and difficult to maintain.

File: web/src/auth/LoginPage.js [hunk 3]
```diff
@@ -380,26 +380,46 @@ class LoginPage extends React.Component {
...
-          </div>
-        </Col>
-      </Row>
-    )
+    let defaultProvider = null;
+    application.providers?.forEach(provider => {
+      if (provider.default) {
+        defaultProvider = provider;
+      }
+    })
```
Comment (sh1luo) [inline web/src/auth/LoginPage.js:388]: same as above

**Review by sh1luo — CHANGES_REQUESTED** [review]

**Comment by sh1luo** [comment]
@WindSpiritSR Forgot to say, except for codes. This demo looks very good.

File: web/src/ApplicationEditPage.js [hunk 4]
```diff
@@ -39,6 +39,7 @@ class ApplicationEditPage extends React.Component {
       application: null,
       organizations: [],
       providers: [],
+      defaultProvider: null,
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:42]: @WindSpiritSR yes, as @sh1luo suggested, remove this `defaultProvider` state.

File: web/src/ApplicationEditPage.js [hunk 5]
```diff
@@ -361,7 +362,7 @@ class ApplicationEditPage extends React.Component {
               this.state.application.enablePassword ? (
                 <SignupPage application={this.state.application} />
               ) : (
-                <LoginPage type={"login"} mode={"signup"} application={this.state.application} />
+                <LoginPage type={"login"} mode={"signup"} application={this.state.application} preview="true" />
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:365]: Change to `isPreview`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/ApplicationEditPage.js [hunk 6]
```diff
@@ -228,6 +228,7 @@ class ApplicationEditPage extends React.Component {
           <Col span={1} >
             <Switch checked={this.state.application.enablePassword} onChange={checked => {
               this.updateApplicationField('enablePassword', checked);
+
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:231]: No this

File: web/src/ApplicationEditPage.js [hunk 7]
```diff
@@ -301,7 +302,8 @@ class ApplicationEditPage extends React.Component {
               table={this.state.application.providers}
               providers={this.state.providers}
               application={this.state.application}
-              onUpdateTable={(value) => { this.updateApplicationField('providers', value)}}
+              onUpdateTable={(value) => {this.updateApplicationField('providers', value)}}
+              enablePassword={this.state.application.enablePassword}
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:306]: Don't pass in enablePassword

File: web/src/ApplicationEditPage.js [hunk 8]
```diff
@@ -373,7 +375,7 @@ class ApplicationEditPage extends React.Component {
           <br/>
           <br/>
           <div style={{width: "90%", border: "1px solid rgb(217,217,217)", boxShadow: "10px 10px 5px #888888",alignItems:'center',overflow:'auto',flexDirection:'column',flex:'auto' }}>
-            <LoginPage type={"login"} mode={"signin"} application={this.state.application} />
+            <LoginPage type={"login"} mode={"signin"} application={this.state.application} isPreview="true" />
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:378]: I don't think `isPreview` is useful, use this.props.application

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**